net: http: Rename `body_start` to `body_frag_start`, implement `body_frag_len`

ycsin opened the pull request:
Following #42026, the `body_start` pointer now points to the start of the body in the `recv_buffer` **unconditionally** as long as there is body in it, either entirely or partially.

This PR renames the `body_start` to `body_frag_start` so the variable now represents what it does better, that it is the start of a body fragment, when body exists either partially or entirely in the `recv_buffer`

The second commit implemented `body_frag_len` which is basically the length of the `body_frag_len` mentioned above, so that user can act on the data directly without dealing with `http_client`'s internals in the application.

```c
                         recv_buffer that contains header + body
                         _______________________________________
            
                                     |←-------- body frag len ---------→|
                  |←--------------------- data len --------------------→|
              ---------------------------------------------------------------
        ..header  |      header      |               body               |  body..
              ---------------------------------------------------------------
                  ↑                  ↑
               recv_buf       body_frag_start


                            recv_buffer that contains body only
                            ___________________________________
        
                   |←------------------ body frag len ------------------→|
                   |←--------------------- data len --------------------→|
              ---------------------------------------------------------------
    ..header/body  |                         body                        |  body..
              ---------------------------------------------------------------
                   ↑
                recv_buf
            body_frag_start

body_frag_start >= recv_buf
body_frag_len = data_len - (body_frag_start - recv_buf)
```

https://github.com/zephyrproject-rtos/zephyr/issues/41683#issuecomment-1018288044

ycsin committed:
net: http: Rename body_start to body_frag_start

Following #42026, the body_start pointer now points to the
start of the body fragment in the recv_buffer as long as there
is body in it, either entirely or partially.

Rename the body_start to body_frag_start to better reflect
what it represents.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

jukkar reviewed (commented):
The `body_len` will not be correct value if we are receiving data in smaller pieces, this can then be really confusing to the end user if he/she trusts this value to really contain the length of the HTTP body. I would avoid this confusion if possible.

ycsin commented:
Tested this just now, not working, changed to draft.

ycsin commented:
> The `body_len` will not be correct value if we are receiving data in smaller pieces, this can then be really confusing to the end user if he/she trusts this value to really contain the length of the HTTP body. I would avoid this confusion if possible.

I'm trying to abstract the `body_len` calculation away from the user of `http_client`, so that a user can get the length of the body data and the pointer to it easily when it exists in the http_response, without going through the source to know what `data_len`, `body_start` & `recv_buf` means respectively.

ycsin commented on subsys/net/lib/http/http_client.c:
```diff
@@ -267,6 +267,8 @@ static int on_body(struct http_parser *parser, const char *at, size_t length)
 	/* Reset the body_start pointer for each fragment. */
 	if (!req->internal.response.body_start) {
 		req->internal.response.body_start = (uint8_t *)at;
+		req->internal.response.body_len = req->internal.response.data_len -
+			(req->internal.response.body_start - req->internal.response.recv_buf);
 	}
```
```suggestion
	}

	/* Calculate the length of the body */
	req->internal.response.body_len = req->internal.response.data_len -
		(req->internal.response.body_start - req->internal.response.recv_buf);
```

jukkar commented:
Note that the `body_start` pointer and possibly new `body_len` is only valid and meaningful if the whole HTTP packet fits to the supplied buffer. For embedded devices, this is probably not the normal case as typically they do not have enough memory for that. So I am not convinced if it makes sense to use the `body_start` and/or `body_len` for anything useful. The `body_start` is only useful if the actual body is starting in currently processed receive buffer and its value should be cleared if we have more data to come. But in that case too, the `body_len` would probably be wrong if it is used to copy data from the receive buffer. The `body_len` value can certainly be used keep track of the total length of the body but that would be needed to document clearly as it does not indicate the amount of bytes in receive buffer.

jukkar reviewed (changes requested):
As I wrote, the `body_len` cannot be calculated like this.

ycsin commented:
I think now I sort of understand why the `body_start` is set only when the `http_client` [first see the start of the body in a http response](https://github.com/zephyrproject-rtos/zephyr/issues/41683#issuecomment-1018288044), so the `body_start` was meant to be pointing to the start of body when a recv buffer is big enough to hold the entire http response? We actually thought that it is a bug and fixed that in #42026, so now the `body_start` would actually points to the start of the body in a `recv_buf`, whether or not the buffer holds the http response partially or entirely.

From my understanding, the `body_start` now means if the `recv_buf` has body in it, and the `body_len` in this PR meant the size of the body for this chunk of `recv_buf`, instead of the actual size of body.

@rlubos thoughts?

ycsin commented on include/net/http_client.h:
```diff
@@ -109,6 +109,9 @@ struct http_response {
 	/** Where the body starts */
 	uint8_t *body_start;
 
+	/** Length of the body data */
```
```c
/** Length of the body contained in the recv_buf */
```

ycsin commented on include/net/http_client.h:
```diff
@@ -109,6 +109,9 @@ struct http_response {
 	/** Where the body starts */
```
```c
/** Start address of the body contained in the recv_buf */
```

jukkar commented:
> I think now I sort of understand why the `body_start` is set only when the `http_client` [first see the start of the body in a http response](https://github.com/zephyrproject-rtos/zephyr/issues/41683#issuecomment-1018288044), so the `body_start` was meant to be pointing to the start of body when a recv buffer is big enough to hold the entire http response?

Not exactly. The `body_start` will only have a valid value when the receive buffer contains the start of body. The whole response does not need to be inside the receive buffer. See https://github.com/zephyrproject-rtos/zephyr/blob/a85099335200758a05dbec207530023cdfd39e1d/subsys/net/lib/http/http_client.c#L447-L455 branch which clears the `body_start` after we have to receive more data and the receive buffer is reused.

So as a summary, if the receive buffer is 1kb, then we only read the data 1kb at a time regardless how long the header / body is. The HTTP parser will call user supplied callback and application needs to act on to the data. So application might receive the data in smaller pieces and need to be prepared for it. Because of this the `body_len` is not a good variable to have and will / could cause confusion. And beside we would know the length of the body only after we have received all the data, and I am not sure how useful that information is at that point.

ycsin commented:
> Not exactly. The body_start will only have a valid value when the receive buffer contains the start of body. The whole response does not need to be inside the receive buffer

Yes you are right. I think we actually changed that behavior in https://github.com/zephyrproject-rtos/zephyr/pull/42026:

> which clears the `body_start` after we have to receive more data and the receive buffer is reused

I believe the remaining body might lead to https://github.com/zephyrproject-rtos/zephyr/blob/a85099335200758a05dbec207530023cdfd39e1d/subsys/net/lib/http/http_parser.c#L2014-L2020

and eventually https://github.com/zephyrproject-rtos/zephyr/blob/a85099335200758a05dbec207530023cdfd39e1d/subsys/net/lib/http/http_client.c#L267-L270

which will set the `body_start` to something again?

ycsin commented:
> The `body_start` will then just tell where in the current receive buffer the body pointer is, it does not tell where the body starts if the message larger than the receive buffer.

Yes and no, `body_start` will tell the real start of the body when the receive buffer first contains it.

> Anyway, the value of `body_len` is limited use as it will just tell how many bytes in this receive buffer contains part of the body. Agreed?

Yes, which I think should be very useful for `http_client`'s user who wants to process that chunk of body data.

> You added some comments but I think this needs to be clarified more in the doxy comments.

Any idea where should I do it?

jukkar commented:
> > The `body_start` will then just tell where in the current receive buffer the body pointer is, it does not tell where the body starts if the message larger than the receive buffer.
> 
> Yes and no, `body_start` will tell the real start of the body when the receive buffer first contains it.

Yes, that is exactly what I said.

rlubos commented:
Perhaps then we should rename the `body_start`/`body_len` to `body_frag_start`/`body_frag_len` to avoid confusion? Having the body pointer set explicitly set for each fragment allows to simplify the actual application logic (see https://github.com/zephyrproject-rtos/zephyr/pull/43237).

jukkar commented:
> Perhaps then we should rename the `body_start`/`body_len` to `body_frag_start`/`body_frag_len` to avoid confusion?

Yes, this looks like a good idea. Also it should be clearly documented in the API what the variables mean, perhaps even with some examples although people usually do not read the docs.

jukkar reviewed (dismissed):
I like the diagram in the PR description, just wondering if it would be possible to place that either to the header file or the commit message. Anyway, LGTM already.

ycsin commented:
> I like the diagram in the PR description, just wondering if it would be possible to place that either to the header file or the commit message. Anyway, LGTM already.

I think it doesnt fit into the commit message, maybe I can place that in the header file, DNM for now

ycsin commented:
- [ ] Edit the OP to reflect the current changes
- [ ] Separate the commit, the first commit should rename the `body_start` to `body_frag_start` and required changes to the `hawkBit`, The following one introduces `body_frag_len`.
- [ ] Paste the diagram to the header

ycsin committed:
net: http: Calculate body_frag_len

Calculate the body_frag_length if any part of the body is found
in the http response.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
@rlubos @jukkar ready for review
